### ISSUE TITLE
How do I get precidential cases?

### ISSUE THREAD

**Issue by karlnicholas** [freelawproject/courtlistener#471]
How do I get cases that are either only precidential or marked on whether they are pecidential or not. With the V3 bulk download I don't see the field anymore and I don't see an obvious filter in the rest api?

Currently primarily interested in california cases.

Thanks.

**Comment by mlissner** [comment]
I haven't gone to double check, but this should be a field on the OpinionCluster object. Have you looked there?

**Comment by ellliottt** [comment]
Yes, it's in OpinionCluster
https://github.com/freelawproject/courtlistener/blob/a6faedb72db0302e9e65e58c49b836d5bd1ebddc/cl/search/models.py#L676

**Comment by karlnicholas** [comment]
Ok, I see it there. I need to as efficiently as possible get full information for a date range of opinions, or at least I need published and case name and date filed from the cluster interface and html-law-box from the opinions interface so I can put everything back together for "complete" opinion information. Should I get a date-range from opinions and then get each of the cluster responses individually or visa-versa?

**Comment by mlissner** [comment]
So you need a join between clusters and opinions between a date range? You
can go either way, but it's probably best to start with the clusters and
then get the opinions because there may be more than one opinion per
cluster.

**Comment by karlnicholas** [comment]
Well, I'm getting back to looking at this. I can get clusters by dates and precedential status, but how do I specify the specific jurisdiction? I am only interested in California Appellate and Supreme court published opinions at this point. I can see "state_cite_one" and state_cite_two" and "state_cite_three" in clusters, but I don't think they are used to specify "court=calctapp" or "court=cal" (or court=cal+calctapp). In fact, I'm not sure I see how I might to that with "opinion" api, so I must need join something else for all this. So, do I need to join the court, opinion, and cluster APIs?

In short, I used to be able to get, from the bulk download, everything I needed, which was the text of the opinions (lawbox), a precedential or published flag, and the opinion dates (filed, create, etc), for all of the California higher courts. I looked at the tables of the schema, but I was able to discern how to do all this via the API.

**Comment by mlissner** [comment]
> Well, I'm getting back to looking at this. I can get clusters by dates and precedential status, but how do I specify the specific jurisdiction?

This might frustrate you, but it gets at the logic of the system. The main two facts to think about are:

1. Information lives at the highest level possible. 
2. Dockets contain OpinionClusters which contain Opinions. 

So, when you're reasoning about where a piece of metadata lives, you should ask yourself: "Could this piece of information logically live on the docket?" If the answer is no, then, "Could this live on the Cluster?" If no: "This must be on the Opinion."

In this case, the jurisdiction of an item pertains to everything that might be in an entire docket, so...the information lives off the docket, and you can filter by jurisdiction using the double underscore syntax:

 - Dockets: https://www.courtlistener.com/api/v3/dockets/?court=calctapp
 - Clusters, joining to dockets: https://www.courtlistener.com/api/v3/clusters/?docket__court=calctapp
 - Opinions, joining to Clusters and onwards to Dockets: https://www.courtlistener.com/api/v3/opinions/?cluster__docket__court=calctapp

Those are off the top of my head, but the idea and the logic should be about right.

> Can I get the opinion text out of the search API?

No, sorry. Indeed, the entire text of the items (with formatting and whatnot) isn't even in the search engine. It lives only in the DB.

**Comment by mlissner** [comment]
I'm assuming the silence here means my lengthy explanation was sufficient. Closing.

**Comment by karlnicholas** [comment]
I was able to use two of the batch files, opinions and clusters, and join them with the URL for the opinion. Actually, I stripped out the ID, converted it to an integer, and used it that way, since that seemed quicker and more reliable. My only suggestion would be to have an ID field that can be used to join opinions and clusters.

**Comment by brianwc** [comment]
Karl, mind sharing the code that worked for you? Might help next person.

**Comment by karlnicholas** [comment]
Well, there's more ins and outs than can be easily posted, but the basic idea is to read the clusters first because it is shorter and put everything into a map with Key=ID and Value=ClusterObject, then read opinions and where ID is found then remove the object from the map, set the HTMLLawBox and OpinionsCited (and/or whatever else you're interested in), and pass it on for parsing.
